src/main/java/com/google/step/servlets/HomePageServlet.java

```diff
@@ -99,60 +104,113 @@ public int compareTo(ScoredDeal other) {
+        && !homePageSection.equals(RESTAURANTS_SECTION)
+        && !homePageSection.equals(TAGS_SECTION)) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+    // if no home page section is being specified to view all deals, return home page data
+    if (userService.isUserLoggedIn()) { // all sections are available
       String email = userService.getCurrentUser().getEmail();
       User user = userManager.readUserByEmail(email);
       long userId = user.id;
-      dealsByUsersFollowed =
+      // Retrieves maps of all the sections
+      List<List<Map<String, Object>>> homePageDealsMaps =
+          getSectionListMaps(homePageSection, userId);
+      if (homePageDealsMaps.size() == 4) {
+        Map<String, Object> homePageMap = new HashMap<>();
+        // for each section, limit to 8 deals for home page
+        homePageMap.put(
+            "popularDeals",
+            homePageDealsMaps.get(0).stream().limit(8).collect(Collectors.toList()));
+        homePageMap.put(
+            "usersIFollow",
+            homePageDealsMaps.get(1).stream().limit(8).collect(Collectors.toList()));
+        homePageMap.put(
+            "restaurantsIFollow",
+            homePageDealsMaps.get(2).stream().limit(8).collect(Collectors.toList()));
+        homePageMap.put(
+            "tagsIFollow", homePageDealsMaps.get(3).stream().limit(8).collect(Collectors.toList()));
+        response.setContentType("application/json;");
+        response.getWriter().println(JsonFormatter.getHomePageJson(homePageMap));
+        // user requested to view all deals of particular section
+      } else if (homePageDealsMaps.size() == 1) {
+        response.setContentType("application/json;");
+        response
+            .getWriter()
+            .println(JsonFormatter.getHomePageSectionJson(homePageDealsMaps.get(0)));
+      }
+    } else {
+      if (homePageSection == null) { // only trending will be shown when not logged in
+        List<List<Map<String, Object>>> homePageDealsMaps =
+            getSectionListMaps(TRENDING_SECTION, -1);
+        response.setContentType("application/json;");
+        response
+            .getWriter()
+            .println(
+                JsonFormatter.getHomePageSectionJson(
+                    homePageDealsMaps.get(0).stream().limit(8).collect(Collectors.toList())));
+      } else if (homePageSection.equals(TRENDING_SECTION)) {
+        // User views all deals for trending section
+        List<List<Map<String, Object>>> homePageDealsMaps = getSectionListMaps(homePageSection, -1);
+        response.setContentType("application/json;");
+        response
+            .getWriter()
+            .println(JsonFormatter.getHomePageSectionJson(homePageDealsMaps.get(0)));
+      } else { // user is unable to view all deals for other sections when not logged in
+        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+        return;
+      }
+    }
```

Comment:
This function is too big. You can probably break it down into 2 different functions (logged in vs logged out)